src/hotspot/share/runtime/vmOperations.cpp

```diff
@@ -279,15 +383,21 @@ void VM_ThreadDump::doit() {
     concurrent_locks.dump_at_safepoint();
   }
 
-  ObjectMonitorsHashtable table;
-  ObjectMonitorsHashtable* tablep = nullptr;
+  ObjectMonitorsDump object_monitors;
   if (_with_locked_monitors) {
-    // The caller wants locked monitor information and that's expensive to gather
-    // when there are a lot of inflated monitors. So we deflate idle monitors and
-    // gather information about owned monitors at the same time.
-    tablep = &table;
-    while (ObjectSynchronizer::deflate_idle_monitors(tablep) > 0) {
-      ; /* empty */
+    // Gather information about owned monitors.
+    ObjectSynchronizer::monitors_iterate(&object_monitors);
+
+    // If there are many object monitors in the system then the above iteration
+    // can start to to take time. Be friendly to following thread dumps by
+    // telling the MonitorDeflationThread to deflate monitors.
+    //
+    // The limit has been arbitrarily chosen to be were the iteration started
+    // to take more then a few milliseconds.
+    size_t monitors_count = ObjectSynchronizer::in_use_list_count();
+    const int DeflateRequestLimit = 100000;
+    if (monitors_count > DeflateRequestLimit) {
+      ObjectSynchronizer::request_deflate_idle_monitors();
```

Comment:
Not sure about this. Arguably, the async deflation policy should re-evaluate the conditions for deflation and then decide to act. Otherwise, this effectively backdoors the heuristics, and does so with the hardcoded threshold. On the other hand, the old code effectively did the same with threshold of `0`. 

So, I would rather keep old behavior and just request deflation without a threshold here.

Reply:
There are a few options how to move forward with this:

1. Stop triggering deflation from the thread dumping code
2. Only trigger if we pass a given limit, say 100000.
3. Always trigger monitor deflation

I believe that long-term it would be best for the JVM if we went with (1). I added (2) just to counter some potential arguments that having too many monitors in the system will make the thread dumping take a long time. It is not clear to me at all that people will notice this, and if they do then maybe we need to tweak the monitor deflation heuristics instead. (3) seems excessive to me.

I've added the flag AsyncMonitorDeflationForThreadDumpLimit. The long name hints that this flag is in support of something overly specific. I set the default to SIZE_MAX in my support for (1), and hope that we can release-note that we have stopped performing monitor deflation from thread dumping and that this flag is going to be good enough safeguard if there are applications that rely on the monitor deflation for thread dumping. But if reviewers disagree with this, I'm OK with changing the default value to support either (2) or (3).

Could I get a all the reviewer's here to (re)state the preference on this, included their suggested limit

I'm also OK with changing the name of the flag if you have a better name, and changing it to an experimental flag if that makes more sense.

Reply:
I would prefer (3), and then consider changing to (1) in a separate PR. This would match current behavior well, and thus would not make more things beyond fixing the interleaving trouble; would eliminate the need to have another flag that would be only temporary until (1) is here; would trigger (pun intended) more discussion about deflation policy once we do (1) -> (3).

Reply:
I didn't get much more feedback on this, so I went with Aleksey's suggestion and ripped out the flag and the ability to set a limit. The code now always requests monitor deflation.